[ky.kiske] 지하철 서비스 미션 코드 리뷰 요청드립니다

SimJung opened the pull request:
잘 부탁드립니다!!! 감사합니다!!

SimJung committed:
feat: 최단거리 경로 구하는 기능 추가

 - Path, PathResult 클래스 추가
 - PathService 서비스 추가
 - Path를 통해 그래프 정보를 초기화하고 메소드를 통해 최단거리 구하도록 하는 기능 추가

SimJung committed:
HandlerMethodArgumentResolver 사용하도록 변경

 - 기존의 Login
Member를 사용하지 않는 방식에서 CustomResolver를 사용하여 LoginMember를 사용하도록 변경

SimJung committed:
feat : Custom Exception 추가

 - InvalidTokenException 클래스 추가
 - LoginFailException 클래스 추가

SimJung committed:
feat : favorite 저장 기능 구현

 - FavoriteService 생성
 - FavoriteDao 생성
 - Favorite Domain 생성
 - Favorite 저장 기능 구현

SimJung committed:
feat : 즐겨찾기 조회 기능 구현 완료

SimJung committed:
feat : 즐겨찾기 삭제 기능 구현 완

SimJung committed:
feat : 노선 추가 요금 정보 조회

 - 노선 추가 요금 데이터 기입
 - 추가 요금 DB 연동

SimJung committed:
feat : 요금 계산 기능 추가

 - step3 요금 계산 기능 추가
 - 거리별, 노선별, 연령별 금액 계산하여 반환

SimJung committed:
refactor : 코드 리팩토링

 - PathService 및 Path for-loop을 stream으로 수정
 - PathResult의 Setter 사용하지 않도록 수정

sehan committed:
chore: 불필요한 lib 파일 제거

gradle dep 으로 의존을 추가했기 때문에 직접 파일을 추가할 필요가 없습니다.

sehan commented:
@SimJung jgrapht-1.5 파일들이 push 되어있어서 리뷰할때 Files changed 가 너무 많아서 제가 직접제거 했습니다. ☺️

sehan commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,46 @@
+        String email = null;
+        if(token != null && jwtTokenProvider.validateToken(token)){
+            email = jwtTokenProvider.getPayload(token);
+        }
```
요청자의 연령대를 알기위해 token 에서 email 을 추출하는 코드를 직접 삽입했네요. 이렇게 집접 inline 코드를 작성하는것 보다는 교육과정에서도 언급된 AuthenticationPrincipalArgumentResolver 를 사용하여 getPaths( @AuthenticationPrincipal xxx ) 방식으로 인증정보를 받아 처리해보는 것을 권합니다. 
@AuthenticationPrincipal 을 사용하게 되면 관련 로직을 분리해 낼 수 있어서 controller 코드의 응집도가 한결 더 개선됩니다. ☺️

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,51 @@
+        this.memberDao = memberDao;
+    }
+
+    public PathResult findShortestPath(Long sourceId, Long targetId, String email) {
```
PathResult 를 만들기 위한 기능이 잘 구현되어 있습니다. 👍 

여기서 Path.findShortestPath() 는 내부에서 사용중인 jgrapht 라이브러리 의존 객체를 return 해주고 있습니다. 라이브러리 의존성은 Path 내에 숨기고 외부로는 제공하지 않는것이 좋습니다. 외부로 노출시킬 수록 jgrapht 에 대한 의존성이 커지게 되어 확장성이 떨어지게 됩니다. Path 의 역할을 명확히 하고 그에 맞는 input, output 타입을 사용해 주세요.

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,51 @@
+        GraphPath result = path.findShortestPath(sourceId, targetId);
+        List<PathVertex> vertexList = result.getVertexList();
+        int distance = (int) result.getWeight();
+        int fare = path.calculateFare(
```
Path 객체는 최단이동경로 와 요금계산 역할을 가지고 있습니다.  두 역할의 응집도가 높지 않아서 역할 분리가 가능해 보입니다.
역할은 분리해 보세요 ☺️

요금계산을 위한 로직을 도메인객체로 이동시켜 주세요

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public PathResult findShortestPath(Long sourceId, Long targetId, String email) {
+        Sections sections = sectionDao.findAll();
```
Data 관점에서 접근을 한다면 sectionDao.findAll() 로 접근하는게 맞습니다. 
Object 관점에서 접근한다면 lineDao.findAll() 하여 전체 노선을 가져와서 최단이동거리를 계산하기 위한 jgrapht 를 초기화 시키는 것이 더 자연스럽습니다. Line 에는 이미 Section 과 Station 객체가 모두 로딩되어 있는 상태입니다. sectionDao.findAll(), lineDao.findAll() 할 필요가 없어집니다. 

sectionDao.findAll() 하게 되면 필요한 데이터를 한번에 가져올 수 있기 때문에 데이터 접근관점에서는 효율적 입니다. 이것은 Data 접근이 필요한 그 부분에 한해서는 아주 올바른 결정입니다. 좀 더 큰 그림을 그려보면 달라 보일 수 있습니다. 최단이동거리를 구하기 위해 request 가 있을때 마다 sectionDao.findAll() , lineDao.findAll() 을 해야할까요? 데이터를 DB 에서 모두 읽어오는건 비용이 큰 작업입니다. 생각해보면 Path 가 가지고 있는 역할은 Application 내에 한번만 로딩되어 @Bean 으로 존재하고 있으면 request 가 있을때마다 DB 에서 데이터를 로딩할 필요없이 최단이동경로를 제공해 주면 됩니다.  

객체(Object) 관점으로 코드를 발전 시켜 주세요 ☺️

sehan commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,116 @@
+        sections.getSections()
+                .forEach(section ->
+                        graph.setEdgeWeight(graph.addEdge(
+                        pathVertices.getPathVertexByStationId(section.getUpStation().getId()),
```
Id 를 넘기는것 보다는 Station 객체를 넘겨서 객체 비교를 통해 찾는것이 더 객체지향적인 방법입니다.

sehan commented on src/main/java/subway/path/domain/PathVertices.java:
```diff
@@ -0,0 +1,50 @@
+        this.pathVertices = pathVertexList;
+    }
+
+    public void initAllVertex(List<Line> lines){
```
객체를 생성하는 방식으로 2가지를 제공하고 있습니다. 
```
// 1
pathV = new PathVertices()
pathV.initAllVertex(lines)

// 2
pathV = new PathVertices(pathVertexList)
```
initAllvertex() 는 2번의 경우에도 호출해서 데이터를 다시 초기화 할 위험이 있습니다. List<Line> 또는 List<PathVertex> 로 객체를 생성하기 위해 두 생성자를 private 로 만들고 static factory method 로 제공을 하는 방법도 있습니다.

sehan commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,56 @@
     // TODO: 즐겨찾기 기능 구현하기
+    @PostMapping(value = "/favorites", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Void> createFavorites(HttpServletRequest request, @RequestBody FavoriteRequest favoriteRequest){
+        String token = AuthorizationExtractor.extract(request);
```
token 으로 사용자의 인증정보(요저정보) 가 필요한 경우 inline 으로 이렇게 작성하지 말고 @AuthenticationPrincipal 어노테이션을사용해서 중복된 inline 코드들을 모두 제거해주세요

sehan commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,56 @@
+import java.util.List;
 
 @RestController
 public class FavoriteController {
```
즐거찾기는 로그인된 (인증을 통과한 또는 인증된 token) 을 가진 요청만 허용합니다. 
interceptor 를 사용하여 접근제한을 구현해야 합니다. 

강의내용을 참고하여 interceptor 를 활용해 보세요 
https://edu.nextstep.camp/s/bb4PXLji/ls/MIOwm2me

sehan reviewed (changes requested):
안녕하세요 카이. 지하철 서비스 리뷰를 맡은 니오 입니다. ☺️
다시 만나서 반가워요~ 😉

최단이동경로찾기를 잘 구현해 주셨습니다. 하지만 인증부분 구현이 누락되어 있습니다.
요금계산기능의 경우 이동경로찾기와 함께 책임과 역할들이 한곳에 몰려 있는 부분들이 보입니다. 
그래서 수정요청 드립니다. 

관련 코멘트를 남겨 두었으니 확인 부탁드려요 ☺️

SimJung committed:
feat : LoginInterceptor 추가

 - 인증 요청을 처리하기 위해 LoginInterceptor 추가
 - 인터셉터 url 정보를 처리하기 위해 config 클래스 추가

SimJung committed:
refactor : inline 코드 변경

 - @AuthenticationPrincipal 어노테이션을 사용하여 loginMember를 받도록 변경

SimJung committed:
refactor : PathController 수정

 - getPaths에서 LoginMember 를 받아 email을 처리하도록 변경

SimJung committed:
feat : PathVertices 생성 방식 추가

 - 기존의 생성자를 이용하여 생성하는 방식에서 static factory method를 통해 생성할 수 있도록 추가 및 수정

SimJung committed:
refactor : 비교 방식 수정

 - 기존의 id를 이용하여 비교하는 방식에서 Station 객체를 비교하는 방식으로 변경

SimJung committed:
refactor : CustomArgumentResolver가 null을 반환하도록 수정

 - 반환된 LoginMember의 값을 바탕으로 컨트롤러에서 핸들링 하도록 수정

SimJung committed:
refactor : findShortestPath 메소드 반환형 수정

 - 기존의 반환형인 라이브러리 의존 객체에서 PathResult를 반환해주도록 수정

SimJung committed:
refactor : fare 클래스 추가

 - Path 객체에서 담당하던 요금 기능을 별도 클래스로 분할
 - PathResponse 를 PathResult와 Fare 객체로 만들 수 있도록 생성자 추가

SimJung committed:
refactor : sectionDao 접근 제거

 - PathService 에서의 불필요한 SectionDao 접근 제거

SimJung committed:
Merge branch 'step1' of https://github.com/SimJung/spring-subway-service-kakao into step1

SimJung commented:
리뷰 감사드립니다!! 덕분에 많은 부분을 개선할 수 있었습니다! 동시에 알아가는 즐거움도 배웠습니다 ㅎㅎ
남겨주신 코멘트를 참고하며 코드를 고치는 와중에 궁금증이 생겨 질문을 드립니다.
1. CustomArgumentResolver 내부에서 오류가 발생하면 오류를 던지고 바깥에서 핸들링을 해야할지, 아니면 그냥 null을 반환해줘야 할지 궁금합니다.
2. Path에서 전체 데이터를 한 번만 읽어오면 이후엔 데이터를 읽어올 필요가 없다는 것으로 이해했습니다. 이럴 경우 Path를 Custom bean 객체로 만들어준다는 의미인지, 이게 맞다면 이후 line 이나 section의 데이터가 변경되어도 자동으로 업데이트가 되는 것인지 궁금합니다..!
감사합니다. 좋은 하루 되세요!!! ☺️

sehan commented on src/main/java/subway/auth/ui/LoginInterceptor.java:
```diff
@@ -0,0 +1,19 @@
+            throw new InvalidTokenException();
+        }
+        return super.preHandle(request, response, handler);
+    }
```
interceptor 를 활용한 접근권한제어 구현을 잘 해주셨습니다. 👍 

여기에 한가지 더 추가하면 더 좋을 내용이 있습니다.  jwt 를 사용하고 있기 때문에 token 에 대한 유효성 검증을 해주면 더 좋습니다.
JwtTokenProvider.validateToken() 을 사용하여 token 검증 로직을 추가해 주세요

sehan commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,55 @@
     // TODO: 즐겨찾기 기능 구현하기
+    @PostMapping(value = "/favorites", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Void> createFavorites(@AuthenticationPrincipal LoginMember loginMember, @RequestBody FavoriteRequest favoriteRequest){
+        Favorite favorite = favoriteService.save(loginMember.getEmail(), favoriteRequest.getSource(), favoriteRequest.getTarget());
```
LoginMember.getId() 로 member 의 Id 를 가져올 수 있습니다. 
favoriteService.save() 시 email 을 넘겨 다시 member 를 조회하는 것 보다는 바로 memberId 를 사용해서 save() 를 하면 됩니다.

sehan commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -24,6 +31,13 @@ public boolean supportsParameter(MethodParameter parameter) {
+        try{
+            String token = AuthorizationExtractor.extract((HttpServletRequest) webRequest.getNativeRequest());
+            Member member = authService.getMemberByToken(token);
+            return new LoginMember(member.getId(), member.getEmail(), member.getAge());
```
요청시 마다 token 으로 DB 에 Member 테이블을 조회해서 LoginMember 객체를 만들 수도 있지만 token 발급시 token 을 발급한 사용자의 LoginMember 객체를 만들어서 관리하면 DB 조회 없이 바로 제공해 줄 수 있습니다. 

이것은 Session 과 유사한 기능입니다. token 에 매핑되는 data 를 유지관리 할 수 있는 기능을 제공해주면 사용자정보(Member) 와 같이 빈번한 DB 요청을 줄여주어 성능개선효과를 누릴 수 있습니다. 

한번 고민해 보고 여력이 된다면 구현 해보세요 😊

sehan commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,42 @@
+    @GetMapping(value = "/paths", params = {"source", "target"}, produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<PathResponse> getPaths(@AuthenticationPrincipal LoginMember loginMember, @RequestParam Long source, @RequestParam Long target){
+        String email = (loginMember == null) ? null : loginMember.getEmail();
+        PathResponse pathResponse = pathService.findShortestPath(source, target, email);
```
member 를 조회하기 위해 많은 곳에서 email 을 사용하고 있네요 😓
최단이동거리 요금을 계산하기 위해서는 age 만 있으면 되겠죠? LoginMember.getAge() 로 얻어 올수 있습니다.

sehan commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -42,20 +51,28 @@ public ResponseEntity createMember(@RequestBody MemberRequest request) {
     @GetMapping("/members/me")
-    public ResponseEntity<MemberResponse> findMemberOfMine(LoginMember loginMember) {
+    public ResponseEntity<MemberResponse> findMemberOfMine(@AuthenticationPrincipal LoginMember loginMember) {
+        if(loginMember == null)
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
요청한 사용자정보 유무를 확인하기 위해 해당 로직이 자주 등장합니다.

이번에 추가한 LoginInterceptor 에 /members/** 관련 path pattern 을 추가해주면 인증된 요청만 controller 에 도달하기 때문에 위 로직을 제거 할 수 있습니다. 
LoginInterceptor 를 활용해서 401 처리하는 방법으로 처리해 보세요 😊

sehan commented on src/main/java/subway/path/domain/Fare.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class Fare {
```
요금계산 역할을 잘 분리했습니다. 👍 👍 👍

sehan commented on src/main/java/subway/path/domain/Fare.java:
```diff
@@ -0,0 +1,62 @@
+    private final int TEEN_UPPER_BOUND = 19;
+
+    public Fare(int distance, List<Integer> extraFareList){
+        fare = calculateFare(distance, extraFareList);
```
생성과 동시에 계산이 진행되는것 보다는 객체 생성자는 생성저 역할을 하고 이미 만들어 놓은 계산매소드로 계산 하도록 하는것이 더 좋은 객체 사용법 입니다.  객체의 모습을 보니 아마도 Fare 를 DTO 로 사용하는 동시에 요금계산 로직을 추가하다보니 그렇게 된거 같아 보이네요.  주객이 뒤바뀐거 같습니다. 😅

Fare = FareCalculator.calculate()  형태가 맞는거 같습니다.  계산과 데이터 전달을 위한 DTO 를 분리하는게 좋겠습니다.

sehan commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,53 @@
+        lines.forEach(line -> path.addSections(line.getSections()));
+
+        PathResult result = path.findShortestPath(stationDao.findById(sourceId), stationDao.findById(targetId));
+        Fare fare = new Fare(result.getDistance(),
+                path.findLineIdListInPath(result.getPathVertices())
+                        .stream()
+                        .map(lineId -> lineDao.findById(lineId).getExtraFare())
+                        .collect(Collectors.toList()));
+
+        if(email != null )
+            fare.discount(memberDao.findByEmail(email).getAge());
```
Fare 로 요금계산 로직을 분리한 것 좋았습니다. 
하지만, 아직도 노선별 추가요금을 위한 로직, 연령별 요금할인 로직이 Service 에 노출되어 있습니다. 
service 에서는 요금 계산객체에게 "path 에 대한 요금을 알려줘" 하는 형태로 끝이 나도록 개선 하면 좋겠네요
아래와 같은 형태의 코드만 남도록 개선 해주세요 😊
```
Fare fare = FareCalculator.calculator();
return new PathResponse(result, fare);
```

요금계산객체 설계시 팁을 드리자면  요금계산에 필요한 요금정책은 3개가 있습니다. 거리별요금정책, 노선별추가요금정책, 연령별요금할인정책.
요금계산시 3가지 정책을 적용하면 최종요금이 결정됩니다.  요금정책을 객체로 만들려면 어떻게 해야 할까요?
만들어진 요금정책객체들이 협력하여 최종요금을 결정하게 하려면 어떻게 해야 할까요? 

요금계산객체는 요금정책을  사용해서 요금을 결정해 주도록 구현 해보시기 바랍니다.

sehan commented on src/main/java/subway/path/domain/Fare.java:
```diff
@@ -0,0 +1,62 @@
+public class Fare {
+    private int fare;
+
+    private final int BASIC_FARE = 1250;
+    private final int FREE_DISTANCE_BOUND = 10;
+    private final int FARE_DISCOUNT_DISTANCE_BOUND = 50;
+
+    private final int KID_LOWER_BOUND = 6;
+    private final int KID_UPPER_BOUND = 13;
+    private final int TEEN_UPPER_BOUND = 19;
```
거리별 요금정책에 관련된 상태와 연령별 할인요금정책에 관련된 상태들이 Fare 에 함께 들어 있습니다. 이 부분을 분리 해 볼 수 있을거 같습니다. ☺️

sehan reviewed (changes requested):
피드백을 잘 반영해 주셨습니다. 👍 👍 👍 

추가로 더 개선하면 좋을 부분들을 코멘트로 남겼습니다.
코멘트 확인 후 개선 부탁드립니다.  

수정요청드립니다. 

추가로, 도메인 객체들 단위테스트를 꼭 작성해 주세요

sehan commented:
> 1. CustomArgumentResolver 내부에서 오류가 발생하면 오류를 던지고 바깥에서 핸들링을 해야할지, 아니면 그냥 null을 반환해줘야 할지 궁금합니다.

Resolver 는 request 에 담긴 요청메세지를 기반으로 argument 를 추출해주는 역할을 합니다. exception 을 던지기 보다는 null 을 반환하는것이 더 적합합니다. 반드시 필요한 필수 argument 라고 한다면 exception 을 던져서 요청메세지 검증역할로 활용할 수도 있지만 일반적으로는 null 을 return 하는 것이 적합해 보입니다.

> 2. Path에서 전체 데이터를 한 번만 읽어오면 이후엔 데이터를 읽어올 필요가 없다는 것으로 이해했습니다. 이럴 경우 Path를 Custom bean 객체로 만들어준다는 의미인지, 이게 맞다면 이후 line 이나 section의 데이터가 변경되어도 자동으로 업데이트가 되는 것인지 궁금합니다..!

Path 를 @Bean 을 사용하여 Bean 으로 등록해서 사용하라는 의도 였습니다.
Bean 으로 등록 후에는 line , section 변경이 발생할 경우 Path 를 갱신 해주는 로직이 필요합니다. 직접 구현해야 합니다. 😊

SimJung committed:
feat : Interceptor 과정 체크 로직 추가

 - jwtTokenProvider의 validation 메소드를 사용하여 토큰 적합성 판단 로직 추가

SimJung committed:
refactor : 불필요한 조회 변경

 - loginMember의 id를 인자로 넘겨 Service에서의 불필요한 dao 조회 없앰

SimJung committed:
refactor : 불필요한 email 조회 삭제

SimJung committed:
refactor : LoginInterceptor 401 에러 핸들링

 - MemberController가 아닌 LoginInterceptor에서 401 에러를 핸들링하도록 수정

SimJung committed:
refactor : Fare 클래스 분할

 - Fare를 Dto 클래스로, FareCalculator 클래스를 util 클래스의 역할을 하도록 분할
 - age의 null 체크를 FareCalculator 에서 하도록 변경

SimJung committed:
feat : Path 생성자 추가

 - List<Line>을 인자로 가지는 Path 생성자 추가

SimJung committed:
refactor : PathVertex 필드 속성 변경

 - 기존의 Line Id를 이용하여 리스트를 관리하는 방식에서 Line 을 이용하여 리스트를 관리하도록 변경

SimJung committed:
refactor : PathService 로직 노출 제거

 - 기존 서비스에서 노출되던 로직 제거 및 기능 이양

SimJung committed:
refactor : 요금 정책 클래스 추가

 - 각 조건에 맞는 요금 정책을 계산하는 클래스를 추가 및 기능 분배

SimJung committed:
feat : LoginInterceptor를Bean으로 관리하도록 변경

 - SubwayConfig를 통해 인터셉터를 Bean으로 관리하도록 변경

SimJung committed:
refactor : Path를 Bean객체로 변환

SimJung committed:
feat : 동일 멤버 조회 시 DB 중복 접근 피하도록 수정

 - Map을 통해 확인하여 이미 존재한다면 해당 value를 return 하도록 수정

SimJung commented:
리뷰 및 코멘트 감사드립니다!! 꼼꼼하게 확인해주셔서 미처 신경쓰지 못했던 부분들을 개선할 수 있었습니다. 정말 감사합니다😊
Path 클래스를 Bean으로 바꾸는 과정에서 코드를 여러번 썼다지웠다 하며 고민해봤습니다만...
Bean으로 사용하게 될 때 초기화를 어떻게 해야할지, 내부에 dao를 autowired로 받아 사용하는게 맞을지 잘 모르겠습니다.
어떤식으로 생각해봐야 할지 고견을 듣고 싶습니다. 
부족한 코드지만 신경써주셔서 감사합니다!!!

SimJung committed:
test : domain Test코드 생성

sehan commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -23,7 +33,18 @@ public boolean supportsParameter(MethodParameter parameter) {
+
+            Member member = authService.getMemberByToken(token);
+            LoginMember loginMember = new LoginMember(member.getId(), member.getEmail(), member.getAge());
+            cache.put(token, loginMember);
```
발급된 token 을 관리하는 로직을 추가해 주셨네요 👍 

해당 역할을 책임질 객체는 누가 더 적합할까? 라는 생각을 더 해본다면 발급된 token 을 관리하는 로직은 token 을 발급하는 곳에 두는것이 좋겠다는 생각이 듭니다.

현재 AuthService 에서 JwtTokenProvider 를 사용하여 발급책임을 가지고 있습니다. 여기서 발급과 동시에 token 을 관리해주면 Resolver 는 argument 를 LoginMember 로 변환해주는 역할에 집중할 수 있고 그외의 로직을 authService 객체로 넘겨 줄 수 있습니다.  또한, AuthService 는 그 역할에 맞게 토큰을 발급하고 토큰의 유효성을 체크하는 것에 집중할 수 있게 됩니다.  public 메소드인 getMemberByToken 은 MemberService 에 더 어울리는 메소드 입니다.

sehan commented on src/main/java/subway/auth/ui/LoginInterceptor.java:
```diff
@@ -0,0 +1,28 @@
+        if(accessToken == null || !jwtTokenProvider.validateToken(accessToken)){
+            response.setStatus(HttpStatus.UNAUTHORIZED.value());
+            return false;
+        }
```
1. JwtTokenProvider 를 주입해서 직접토큰 검증을 수행 할 수도 있지만, token 을 발급해주는 AuthService 에서 발급한 token 검증을 해주는것이 응집력 측면에서 더 좋아 보입니다.
AuthService 를 주입받아 authService.validateToken() 해주는 건 어떨까요?

2. 참고 ). 위 코드처럼 token 이 유효하지 않은경우 401 응답처리를 하기 위해 response 를 각 상황에 따라 처리하는 것도 방법입니다. 
또 다른 방법으로는 각 상황별로 response 를 처리하는 것이 아니라 response 에 대한 에러처리를  한곳에서 담당하게 하는 방법도 있습니다.  spring 에서 에러처리하는 방법들을 소개한 글 한번 살펴보세요 
https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc
https://www.baeldung.com/exception-handling-for-rest-with-spring

sehan commented on src/main/java/subway/config/SubwayConfig.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(loginInterceptor()).addPathPatterns("/favorites/**");
+        registry.addInterceptor(loginInterceptor()).addPathPatterns("/members/me");
```
아래와 같이 다수의 pattern 을 설정하면 됩니다.
```suggestion
        registry.addInterceptor(loginInterceptor()).addPathPatterns("/favorites/**", "members/me");
```

sehan commented on src/main/java/subway/path/domain/FareCalculator.java:
```diff
@@ -0,0 +1,47 @@
+        FareByAge fareDiscount = new FareByAge(basicFare.getFare() + extraFare.getFare(), age);
+
+        return fareDiscount;
+    }
```
FareByDistance, FareByLine, FareByAge 가 FareCalculator 의 멤버라면 calculate 마다 객체를 생성하지 않아도 될거 같네요.
이렇게 되면 calculate() 매소드에서 static 을 제거할 수 있고 그에 따라 어쩔수 없이 static 으로 선언한 private 매소드들 에서도 static 을 제거 할 수 있습니다. 
꼭 static 이어야 하는게 아니라면 static 을 제거 하는것이 객체를 더 객체 답게 만들어 줍니다. ☺️
static 매소드를 작성할때 그에 부가적으로 private static 매소드들을 만들어야 하는일이 발생하게 되면 static 매소드가 맞는가? 하는 의심을 해보길 바랍니다.

sehan commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,54 @@
+    private PathVertices pathVertices;
+
+    @Autowired
+    private LineDao lineDao;
```
도메인 객체에 Spring 어노테이션을 사용하는것을 도메인이 Spring 에 의존하게 만듭니다.
도메인은 spring 이 아닌 다른 프레임웍이나 다른 환경에서도 사용할 수 있는것이 재사용성, 확장성 측면에서 좋습니다.

생성자에서 List<Line> 을 받아 Path 를 생성하는것이 좋습니다.
LineDao 의존을 제거 해주세요

sehan commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,54 @@
+                                pathVertices.getPathVertexByStation(section.getUpStation()),
+                                pathVertices.getPathVertexByStation(section.getDownStation())),
+                                section.getDistance()));
+    }
```
PathVertex 에 List<Line> 을 추가하여 최단이동경로의 요금계산시 노선추가요금계산에 사용하는데 이로 인해 로직들이 많이 복잡해 진것이 아닌가 생각이 듭니다. 

PathVertex 를 만들어서 jgraph 에 vertex 로 추가해주는데 PathVertex 는 jgraph 에만 저장되기 때문에 Path 에서도 pathVertices 상태를 관리하게 됩니다. jgraph 의 vertex 를 구성할때 vertex 로서의 정보이외에 요금계산을 위한 데이터를 추가하여 노선추가요금 계산로직에 활용하다보니 로직이 복잡해 졌습니다. 

이를 개선하기 위해서는 노선추가요금계산 로직을 위해 추가한 List<Line> 을 PathVertex 에서 제거하고 노선추가요금은 구간정보를 입력하면 노선추가요금을 알려주는 요금테이블 데이터를 구성해 놓고 Path.findShortestPath() 의 결과를 가지고 노선추가요금을 계산하는것이 로직의 복잡도를 낮춰 줄거 같습니다.

sehan commented on src/main/java/subway/config/SubwayConfig.java:
```diff
@@ -0,0 +1,28 @@
+    @Bean
+    public Path path(){
+        return new Path();
+    }
```
Path를 생성할때 List<Line> 을 생성인자로 받아서 생성하면 LineDao 의존을 제거 할 수 있습니다. 
```
public Path path(LineDao lineDao){
    return new Path(lineDao.findAll())
}
```

Path 는 도메인 객체이니 단위테스트를 작성해 보시면 어떤 의미인지 바로 알수 있습니다. ☺️
DB 의존때문에 단위테스트 작성이 쉽지 않을 것입니다

sehan commented on src/test/java/subway/path/domain/FareByAgeTest.java:
```diff
@@ -0,0 +1,21 @@
+        assertThat(baby.getFare()).isEqualTo(0);
+        assertThat(young.getFare()).isEqualTo(1675);
+        assertThat(teen.getFare()).isEqualTo(2470);
+        assertThat(adult.getFare()).isEqualTo(3000);
```
요구사항에 맞게 단위테스트를 나누어서 작성해주는것이 좋습니다. 
단위테스트를 보는 것이 기능명세를 작성한다는 느낌으로 작성해주면 좋습니다. 

단위테스트에 하나의 assert 만 넣어서 작성해 보세요

sehan commented on src/test/java/subway/path/domain/FareByDistanceTest.java:
```diff
@@ -0,0 +1,14 @@
+    void distanceTest(){
+        FareByDistance fare = new FareByDistance(30);
+        assertThat(fare.getFare()).isEqualTo(1650);
+    }
```
거리별 요금정책을 단위테스트로 명세한다는 느낌으로 단위테스트를 추가로 작성해 주세요. 

10km 미만, 10~50km, 50km 추가시 각 상황에 맞게 결과가 나오는지 작성해 주세요

sehan commented on src/test/java/subway/path/domain/FareByLineTest.java:
```diff
@@ -0,0 +1,16 @@
+    @Test
+    void lineTest(){
+        FareByLine fare = new FareByLine(Arrays.asList(new Line("l1", "c1", 500), new Line("l2", "c2", 1000)));
+        assertThat(fare.getFare()).isEqualTo(1000);
```
FareXXX 객체들은 모두 생성과 함께 로직이 실행되고 요금을 제공하는 DTO 역할도 수행하고 있습니다. 
DTO 와 같은 객체가 요금계산 비지니스를 가지고 있습니다. 그 부분이 테스트 코드를 보니 확실히 드러나고 있습니다. 

노선별 추가요금을 계산하는 객체라면 단위테스트에서 아래 같은 형태로 요금을 계산하는 코드 전개가 나올거 같습니다. 
```
policy = LineExtraFarePolicy( 노선구간별추가요금테이블 ) 
노선추가요금 = policy.apply( 최단이동경로 )
```

sehan commented on src/test/java/subway/path/domain/PathTest.java:
```diff
@@ -0,0 +1,8 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+class PathTest {
+}
```
중요한 findShortestPath() 에 대한 테스트가 없습니다. 😅
추가해 주세요

sehan commented on src/test/java/subway/path/domain/PathVerticesTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    void getTest(){
+        assertThat(pathVertices.getPathVertexList()).isNotEmpty();
+        assertThat(pathVertices.getPathVertexByStation(s1)).isNotNull();
```
assertThat(pathVertices.getPathVertexList()).isNotEmpty();
not empty 체크 뿐 아니라 getPathVertexList() 에서 return 되는 List 가 vertex 를 정말 포함하고 있는지를 체크하는 것이 좋습니다.  

assertThat(pathVertices.getPathVertexByStation(s1)).isNotNull();
여기에서도 not null 뿐 아니라 실제로 기대했던 PathVertex 가 나왔는지 확인해야 합니다.

sehan reviewed (changes requested):
피드백 반영을 위해 많이 고심하셨네요 😊

요금계산역할을 객체로 잘 분리해 주셨습니다.
도메인 단위테스트 작성까지 잘 해주셨는데 아쉬운 부분들이 있어서 한번더 수정요청 드립니다.

자세한 내용은 코멘트를 확인해 주세요.

SimJung committed:
refactor : 토큰 관리 AuthService로 기능 이전

 - 기존의 ArgumentResolver에서 관리하던 토큰 관리 맵을 AuthService로 이전
 - AuthService의 메소드인 getMemberByToken을 MemberService로 이동
 - 로그인 인터셉터의 validateToken을 AuthService를 통해 확인하도록 변경

SimJung committed:
refactor : Exception에 HttpStatus 어노테이션 추가

 - InvalidTokenException에 어노테이션을 통한 UNAUTHORIZATION 상태 코드 추가

SimJung committed:
refactor : 인터셉터 path 설정 간소화

SimJung committed:
refactor : FareCalculator 리팩토링

 - calculate 메소드 및 타 메소드 static 선언부 제거
 - 계산에 필요한 요금 객체들을 필드로 지정
 - 객체들의 calculateFare 메소드의 반환형을 변경

SimJung committed:
fix : 연령별 요금 처리 로직 픽스

SimJung committed:
test : 테스트 코드 리팩토링

 - 테스트를 통해 기능 및 결과를 쉽게 알 수 있도록 수정
 - DisplayName 어노테이션 추가
 - 테스트 별 assert 구문 분리

SimJung committed:
refactor : PathVertex 로직 변경

 - List<Line>을 제거
 - PathVertices의 static factory method의 인자를 기존의 List<Line> 에서 List<Station>을 대신 사용

SimJung committed:
feat : Sections 도메인 메소드 추가

 - 일치하는 역들을 이용하여 섹션을 검색하는 메소드 추가

SimJung committed:
refactor : 노선 추가요금 계산 로직 분리

 - PathVertices 도메인에 최단 루트를 거쳐온 노선 리스트를 반환하는 메소드를 만들어줌.
 - 지나는 노선의 정보를 Service에서 만든 PathResult를 바탕으로 만들어줌.
 - FareCalculator에서 해당 노선 목록을 인자로 받아 계산하도록 수정

SimJung committed:
refactor : 요금 계산 클래스 변경

 - DTO의 역할을 제외하고 요금계산 로직만을 가지도록 이름 및 기능을 수정

SimJung committed:
refactor : PathVertex Equals 추가

 - Station 값을 통해 비교하기 위해 Equals 와 hashCode 오버라이딩
 - PathVerticesTest 에서 실제로 해당 값을 가지고 있는지 확인하도록 변경

SimJung committed:
refactor : Path 클래스 리팩토링

 - LineDao 의존 삭제
 - Configuration에서 list를 통해 생성될 수 있도록 설정
 - PathTest 테스트 코드 작성

SimJung committed:
feat : Line, Section 도메인 메소드 추가

 - 해당 Section을 Line, 혹은 Sections에 가지고 있는지 반환하는 메소드 생성
 - 일치 여부를 판단하기 위해 Section의 equals와 hashCode 메소드 오버라이딩

SimJung committed:
refactor : PathVertices 리팩토링

 - 추가요금 리스트를 구하는 로직의 메소드를 분리
 - 해당 메소드 테스트 완료

SimJung commented:
지난 코멘트와 리뷰 정말 감사드립니다!! 꼼꼼하게 짚어주신 부분들을 통해 배울 수 있었습니다. 감사합니다!!
말씀해주신 부분들을 바탕으로 코드를 개선해보았는데 Path 부분에서 여전히 문제가 있는 것 같아 질문 드립니다.
Config 파일에서 List를 받아 생성해주는 부분으로 변경하고 진행하다보니 비어있는 데이터가 들어오는 것을 확인했습니다.
처음부터 로드를 한 번만 하고 이후로는 추가적인 DB 조회 없이 bean으로 생성된 path 객체를 통해 최적 경로를 계산하고 싶었습니다만, 생각대로 잘 되지 않아 일단은 진행을 위해 initPath 메소드를 둔 상태입니다. 여러번 주신 힌트에도 불구하고 아직 감이 안 잡히는 것 같습니다...😭 
subwayConfig 파일에서 설정을 해줘야하는 것인지, 아니면 추가적인 어떤 과정이 필요한 것인지 알고 싶습니다.
부족한 코드지만 리뷰해주셔서 감사합니다!!!

sehan commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,56 @@
+        this.jwtTokenProvider = jwtTokenProvider;
+    }
+
+    public TokenResponse createToken(TokenRequest tokenRequest){
+        if(!checkEmailValidation(tokenRequest.getEmail(), tokenRequest.getPassword())){
+            throw new LoginFailException();
+        }
+        String accessToken = jwtTokenProvider.createToken(tokenRequest.getEmail());
+        return new TokenResponse(accessToken);
+    }
```
token 발급한다는 것은 해당 token 을 포함하는 요청의 경우 사용자 식별이 가능하다는 의미이기도 합니다.
사용자 식별이 되면 사용자ID 가 없어도 사용자 식별이 가능하기 때문에 접근권한체크도 가능하고 사용자정보도 매핑시킬 수 있습니다.

사용자정보의 경우 요청시 마다 자주 사용되는 항목이기 때문에 token 을 사용하여 DB 조회 없이 정보를 제공해주면 DB 조회를 줄여주는 효과를 얻을 수 있어서 token - context(사용자정보) 를 보관(cache) 해두면 좋습니다.  cache 를 멤버를 추가해서 이 부분은 잘 구현해 주셨습니다. 

cache 에 사용자정보(loginMember) 가 저장되는 시점은 token 이 발급됨과 동시에 이루어지는 것이 가장 이상적인 위치 입니다. 외부에서 putLoginMember() 사용해서 cache 에 넣어주는것보다는 AuthService 가 그 책임을 가지고 내부적으로 처리하는 것이 더 좋습니다. 그렇지 않으면 외부에서 cache 를 위해 member 객체를 만들어서 넣어줘야 하는 것을 알아야 하기 때문입니다. 
```
String accessToken = jwtTokenProvider.createToken(tokenRequest.getEmail());
cache.put(accessToken, memberDao.findByEmail(tokenRequest.getEmail());
```

이렇게 되면, token 이 invalid 한 상황이 왔을때는 cache 를 삭제하는 로직도 필요하게 됩니다. ☺️
점점 구현해야 할 사항들이 늘어나죠.  모든것들 완벽히 구현하는 것 보다는 이런 것들이 필요하겠구나? 하는 생각에 도달하기를 의도 했습니다. 고민과 깨달음을 얻은 시간이 되었기를 바랍니다. 😊

추가, 그리하여 MemberService 에 AuthService 의존을 추가하여 .getMemberByToken() 메소드를 추가한것은 상당히 부적절한 조치 입니다 😄   AuthService 에서 memberDao 를 이용하여 필요한 가져와 cache 에 넣어두고  새로운 요청에 token 이 포함되어 있는 경우 AuthenticationPrincipalArgumentResolver 가 authService 에게 token 의 context 를 요청해서(DB조회없이) 얻은 loginMember 를 controller 에 제공( return ) 해주면 됩니다.

SimJung replied:
아하... 뭔가 로직이 이상하게 늘어간다 싶으면 각 역할에 충실하지 못한 설계 덕분인 것 같습니다😂 감사합니다!!

sehan commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,61 @@
+        return new PathResult(PathVertices.of(result.getVertexList()), (int)result.getWeight());
+    }
+
+    public List<Line> getLines() {
+        return lines;
+    }
```
노선별 추가요금을 계산하기 위해 Path.getLines() 를 사용해서 전체노선을 가져오고 있는데 이부분도 Path 가 제공해줄 필요는 없어 보입니다. PathService.findShortestPath() 에서 이미 lindDao.findAll() 해서 전체노선을 가져왔기 때문에 Path 에서 꺼낼 필요 없이 바로 사용하면 해결이 됩니다. 
 
FareCalculator.calculate() 시 노선추가요금계산에 필요한 데이터를 만들기 위해 최단이동경로가 이용한 노선추가요금 목록을 만들어서 넘겨주었습니다.  이렇게 된 원인을 살펴보니 LineExtraFarePolicy 가 하는역할이 List<Integer> 를 받아 max 값을 구하는 것 밖에 없습니다. 노선추가요금에 필요한 로직들이 외부에 존재하고 내부에서는 max 값 구하는 로직만 있기 때문입니다.  노선추가요금을 계산하는 비지니스로직들이 LineExtraFarePolicy 로 이동하면 더 좋겠습니다.

SimJung replied:
같은 도메인이라도 어떤 도메인에 추가하는지도 중요하군요!!! 코멘트 감사합니다!!

sehan reviewed (approved):
피드백 반영 잘 해주셨습니다. 👍 

요금계산 관련 리팩토링을 위해 노력을 많이 해주셨습니다. 
단위테스트 작성도 아주 잘 되었구요. 코드도 이전보다 많이 정리가 되었습니다. 👍 👍 👍 
요금정책내부 구현시 사용된 분기문을 제거 해보는 시도를 해본다면 더 개선된 코드를 만나 보실 수 있습니다. ☺️

피드백 잘 따라와 주셔서 고맙고 감사합니다. 
남은 마지막 미션 잘 마무리하고 부서로 오는날 만나요 😊

merge 합니다.

sehan commented:
> 말씀해주신 부분들을 바탕으로 코드를 개선해보았는데 Path 부분에서 여전히 문제가 있는 것 같아 질문 드립니다.
> Config 파일에서 List를 받아 생성해주는 부분으로 변경하고 진행하다보니 비어있는 데이터가 들어오는 것을 확인했습니다.
> 처음부터 로드를 한 번만 하고 이후로는 추가적인 DB 조회 없이 bean으로 생성된 path 객체를 통해 최적 경로를 계산하고 싶었습니다만, 생각대로 잘 되지 않아 일단은 진행을 위해 initPath 메소드를 둔 상태입니다. 여러번 주신 힌트에도 불구하고 아직 감이 안 잡히는 것 같습니다...😭
> subwayConfig 파일에서 설정을 해줘야하는 것인지, 아니면 추가적인 어떤 과정이 필요한 것인지 알고 싶습니다.

네 맞습니다. 아직 혼돈에 빠져있는거 같습니다. 😅
```
   @Bean
    public Path path(LineDao lineDao){
        return new Path(lineDao.findAll());
# 전체노선만 생성자로 넘겨주고 생성자 내부에서 jgraph 를 만들어주는 initPath() 호출부분이 없네요
    }
```
